Core/src/org/sleuthkit/autopsy/geolocation/datamodel/WaypointBuilder.java

```diff
@@ -619,44 +622,88 @@ static private String getWaypointListQuery(List<DataSource> dataSources) {
                 dataSourceList);
     }
 
+    /**
+     * A parser that could throw a GeoLocationDataException when there is a
+     * parse issue.
+     *
+     * @param <T> The return type.
+     */
+    private interface ParserWithError<T> {
+
+        T parse(BlackboardArtifact artifact) throws GeoLocationDataException;
+    }
+
+    /**
+     * Parses one waypoint.
+     *
+     * @param parser The parser to use.
+     * @param artifact The artifact to be parsed.
+     *
+     * @return Returns a parse result that is either successful with a parsed
+     *         waypoint or unsuccessful with an exception.
+     */
+    private static GeoLocationParseResult<Waypoint> parseWaypoint(ParserWithError<Waypoint> parser, BlackboardArtifact artifact) {
+        try {
+            return GeoLocationParseResult.create(artifact, parser.parse(artifact));
+        } catch (GeoLocationDataException ex) {
+            return GeoLocationParseResult.error(artifact, ex);
+        }
+    }
+
+    /**
+     * Parses a list of waypoints.
+     *
+     * @param parser The parser to use.
+     * @param artifact The artifact to be parsed.
+     *
+     * @return Returns a list of parse results with the successfully parsed
+     *         items or a list with one item indicating the failure.
+     */
+    private static List<GeoLocationParseResult<Waypoint>> parseWaypoints(ParserWithError<List<Waypoint>> parser, BlackboardArtifact artifact) {
+        try {
+            return parser.parse(artifact).stream()
+                    .map((result) -> GeoLocationParseResult.create(artifact, result))
+                    .collect(Collectors.toList());
+
+        } catch (GeoLocationDataException ex) {
+            return Arrays.asList(GeoLocationParseResult.error(artifact, ex));
+        }
+    }
+
     /**
      * Create a Waypoint object for the given Blackboard artifact.
      *
      * @param artifact The artifact to create the waypoint from
      * @param type     The type of artifact
      *
      * @return A new waypoint object
-     *
-     * @throws GeoLocationDataException
      */
-    static private List<Waypoint> getWaypointForArtifact(BlackboardArtifact artifact, ARTIFACT_TYPE type) throws GeoLocationDataException {
-        List<Waypoint> waypoints = new ArrayList<>();
+    private static List<GeoLocationParseResult<Waypoint>> getWaypointForArtifact(BlackboardArtifact artifact, ARTIFACT_TYPE type) {
+        List<GeoLocationParseResult<Waypoint>> waypoints = new ArrayList<>();
         switch (type) {
             case TSK_METADATA_EXIF:
-                waypoints.add(new EXIFWaypoint(artifact));
+                waypoints.add(parseWaypoint((a) -> new EXIFWaypoint(a), artifact));
```

Comment:
This is very confusing. Could you instead go for an API like parseWaypoint(artifact, EXIFWaypoint.class)?

Reply:
How about something like `waypoints.add(parseWaypoint(EXIFWaypoint::new, artifact));`?